### PR TITLE
Configure Code Climate integrations

### PR DESCRIPTION
**Pull request by dblandin**
> Which issue, if any, is this issue related to?

None, but it's a follow-up of #2062

> Is there anything in the PR that needs further explanation?

Hi there! Devon from @codeclimate here. 👋 

First off, the conversations in #2062, #1341, and https://github.com/codeclimate/codeclimate-eslint/issues/80 have been incredibly helpful internally as we've worked to improve the platform and build features. Thanks for the discussion!

I thought I'd take another go at configuring Code Climate analysis for stylelint now that we've worked out a few of the issues you brought up before.

The bulk of this PR includes commits to craft a Code Climate configuration to fit your codebase. Each commit has a bit more detail under the message fold.  I've enabled and configured our engines for [ESLint][engine-eslint], [duplication][engine-duplication], [Node Security][engine-nsp], and [MarkdownLint][engine-markdownlint].

One blocker in the past has been a lack of support for external shared configurations. In November of last year we [released a feature](https://codeclimate.com/changelog/582495c32c33066f1b00191d) which customers can use to download shared configurations into the project workspace prior to analysis. I've configured the `prepare` block in the Code Climate configuration file to take advantage of this feature and pull down [your shared ESLint configuration](https://github.com/stylelint/eslint-config-stylelint).

While configuring Code Climate, I've done my best to not interfere with any local invocations of ESLint or other dev tools. The changes proposed in this PR should only affect Code Climate and CI environments.

I've forked your repository under our testing organization so that I could simulate active integrations:

Master branch: https://github.com/codeclimate-testing/stylelint
PR introducing new issues: https://github.com/codeclimate-testing/stylelint/pull/3
Code Climate PR page: https://codeclimate.com/github/codeclimate-testing/stylelint/pull/3
Code Climate Repo page: https://codeclimate.com/github/codeclimate-testing/stylelint

I've added screenshots of GitHub augmented by our Chrome extension to the simulation PR. If you'd like to try out the extension for youself, you can do so here: https://codeclimate.com/browser-extension.

Cheers!

cc/ @davidtheclark, @ntwb 

[engine-eslint]: https://github.com/codeclimate/codeclimate-eslint
[engine-duplication]: https://github.com/codeclimate/codeclimate-duplication
[engine-nsp]: https://github.com/nodesecurity/nsp
[engine-markdownlint]: https://github.com/codeclimate/codeclimate-markdownlint

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
...
+[![NPM version](https://img.shields.io/npm/v/stylelint.svg)](https://www.npmjs.org/package/stylelint)
+[![Build Status](https://travis-ci.org/stylelint/stylelint.svg?branch=master)](https://travis-ci.org/stylelint/stylelint)
+[![Build status](https://ci.appveyor.com/api/projects/status/wwajr0886e00g8je/branch/master?svg=true)](https://ci.appveyor.com/project/stylelint/stylelint/branch/master)
+[![NPM Downloads](https://img.shields.io/npm/dm/stylelint.svg)](https://www.npmjs.org/package/stylelint)
+[![Bountysource](https://www.bountysource.com/badge/tracker?tracker_id=9282518)](https://www.bountysource.com/trackers/9282518-stylelint?utm_source=9282518&utm_medium=shield&utm_campaign=TRACKER_BADGE)
+[![Code Climate](https://lima.codeclimate.com/repos/588fd23246e87530aa00185f/badges/af300ac24dac8d18dc60/gpa.svg)](https://lima.codeclimate.com/repos/588fd23246e87530aa00185f/feed)
```
Comment (ntwb) [inline README.md]: Both these URLs result in 404 errors?

File: README.md [hunk 2]
```diff
@@ -1,6 +1,12 @@
...
+[![Test Coverage](https://lima.codeclimate.com/repos/588fd23246e87530aa00185f/badges/af300ac24dac8d18dc60/coverage.svg)](https://lima.codeclimate.com/repos/588fd23246e87530aa00185f/coverage)
```
Comment (ntwb) [inline README.md]: Also both these URLs result in 404 errors?

File: package.json [hunk 3]
```diff
@@ -94,10 +94,11 @@
   "scripts": {
     "benchmark-rule": "node scripts/benchmark-rule.js",
     "flow": "flow",
-    "jest": "jest",
+    "jest": "jest --coverage",
     "lint:js": "eslint . --cache",
     "lint:md": "remark . --quiet --frail",
     "lint": "npm-run-all --parallel lint:*",
+    "preci": "npm-run-all --serial lint:md flow",
```
Comment (ntwb) [inline package.json]: Why is this required? Why not use the existing `pretest` considering you're pulling in `eslint-config-stylelint`:

> _"I've configured the prepare block in the Code Climate configuration file to take advantage of this feature and pull down your shared ESLint configuration."_

File: package.json [hunk 4]
```diff
@@ -94,10 +94,11 @@
   "scripts": {
     "benchmark-rule": "node scripts/benchmark-rule.js",
     "flow": "flow",
-    "jest": "jest",
+    "jest": "jest --coverage",
```
Comment (ntwb) [inline package.json]: Even though this PR is currently passing the Travis CI tests I think this is going to be an issue per #2174 

_FYI: Once I've finished reviewing this PR I'll take another look at #2174 and add another note here_

File: .travis.yml [hunk 5]
```diff
@@ -6,20 +9,26 @@ cache:
...
   include:
     - node_js: '7'
-      script: npm run pretest
-      env: CI=pretest
+      script: npm run preci
+      env: CI=preci
     - node_js: '7'
-      script: npm run jest -- --runInBand
-      env: CI=tests 7
+      script: npm run jest -- --coverage --runInBand
```
Comment (ntwb) [inline .travis.yml]: Even though this PR is currently passing the Travis CI tests I think this is going to be an issue per #2174 

_FYI: Once I've finished reviewing this PR I'll take another look at #2174 and add another note here_

File: .travis.yml [hunk 6]
```diff
@@ -1,3 +1,6 @@
+addons:
+  code_climate:
+    repo_token: 2983486985576ac8d53c4716ab69217e6bad501f42f69138cc796fda85637a6e
```
Comment (ntwb) [inline .travis.yml]: Is this token generated from your GitHub repo fork or a token generated by code climate?

File: package.json (hunk 3 above)
Comment (ntwb) [inline package.json]: Does this have something to do with the fact you are configuring ESLint and markdownlint in the `.codeclimate.yml` file and that we use remarklint rather than markdownlint?

File: .travis.yml [hunk 7]
```diff
@@ -6,20 +9,26 @@ cache:
 matrix:
   include:
     - node_js: '7'
-      script: npm run pretest
-      env: CI=pretest
+      script: npm run preci
+      env: CI=preci
```
Comment (ntwb) [inline .travis.yml]: By making the above change of removing the `pretest` Travis CI job in favour of your `preci` job you've removed the only job that ran ESLint as part of any Travis CI job.

**Comment by ntwb** [comment]
I'm not sure why the commit https://github.com/stylelint/stylelint/pull/2310/commits/964de37b8f112482c1c234351347db3c69e47b37 makes any changes the `npm-shrinkwrap.json` file, it shouldn't.

**Comment by ntwb** [comment]
> Hi there! Devon from @codeclimate here. 👋

Thanks for creating the PR Devon, I've got a bit to say here in regard to this PR beyond what I've initially added via inline comments, I'll start adding those thoughts in follow up comments here shortly.

**Comment by ntwb** [comment]
> > First off, the conversations in #2062, #1341, and codeclimate/codeclimate-eslint#80 have been incredibly helpful internally as we've worked to improve the platform and build features. Thanks for the discussion!

Per https://github.com/stylelint/stylelint/pull/2062#issuecomment-260961709
> Yes, but the ones I looked at were not actionable. They were for code that _should_ be duplicated, for which an abstraction would be overkill. But, yeah, I'm sure there are some cases where this is not the case.

This still appears to be the case when I look at https://codeclimate.com/github/codeclimate-testing/stylelint/issues

Per #1341 Create Code Climate engine

This PR is only to facilitate code analysis of stylelint and *not* the actual creation of a code climate stylelint engine correct?

Per codeclimate/codeclimate-eslint#80

It's nice that you've now added support for repos to _fetch_ their own ESLint configurations.

**Comment by ntwb** [comment]
>  I've enabled and configured our engines for ESLint, duplication, Node Security, and MarkdownLint.

ESLint:
• This pull request has removed ESLint running as part of any Travis CI job.

• There are numerous false positives reported by ESLint here as you're pulling in `master` of `eslint-config-stylelint` rather than version `6.0.0` per our `package.json`

Markdown:

• The markdownlint rules you've added do not match our remarklint configuration and causes many false positives by code climate https://codeclimate.com/github/codeclimate-testing/stylelint/issues

• Unless markdownlint can match verbatim our remarklint configuration, or there is a code climate remarklint engine I think markdownlint should be removed.

Duplication: 

• I noted this above https://github.com/stylelint/stylelint/pull/2310#issuecomment-276561755, so maybe _duplication_ should also be removed

Node Security:

• It doesn't appear that this is available to use as a filter when looking at the issues https://codeclimate.com/github/codeclimate-testing/stylelint/issues, does this not show if there are no security issues or is it a code climate bug? (p.s. I just checked and there are no outstanding NSP issues with stylelint)

**Comment by ntwb** [comment]
I think that's all I've got for now, apart from a question for @stylelint/core: and @dblandin:

Per my above comments what benefit does analysis of stylelint by code climate bring us?

In the simplest of terms I see this PR is to really only add a _code climate badge_ to the stylelint repo, the only enhancement I see is the addition of _node security_, the _duplication_ and _markdownlint_ engines reports are inaccurate.

An interesting side effect of this PR is that `jest -coverage` now works, which I'm presuming would now allow us to resolve #2174

We have a few choices here:

1. Use code climate as our "tool" of choice for reporting Jest' code coverage analysis
2. Use coveralls as our "tool" of choice for reporting Jest' code coverage analysis
3. Use neither and only run coverage locally per https://github.com/stylelint/stylelint/pull/2174#issuecomment-265550584

**Comment by dblandin** [comment]
> I'm not sure why the commit 964de37 makes any changes the npm-shrinkwrap.json file, it shouldn't.

The nodesecurity Code Climate engine runs in offline mode, requiring an `npm-shrinkwrap.json` file in the repository. If having that file in the repository is undesirable, I can disable the engine.

---

> This PR is only to facilitate code analysis of stylelint and not the actual creation of a code climate stylelint engine correct?

Yep, that's correct.

---

**ESLint**

> • This pull request has removed ESLint running as part of any Travis CI job.

That's correct. This PR suggests running ESLint via Code Climate instead.

> • There are numerous false positives reported by ESLint here as you're pulling in master of eslint-config-stylelint rather than version 6.0.0 per our package.json

I've updated the prepare definition in `.codeclimate.yml` to pin to `v6.0.0` of your shared configuration.

---

**Markdown**

> 
> • The markdownlint rules you've added do not match our remarklint configuration and causes many false positives by code climate https://codeclimate.com/github/codeclimate-testing/stylelint/issues
>
> • Unless markdownlint can match verbatim our remarklint configuration, or there is a code climate remarklint engine I think markdownlint should be removed.

Agreed. There might be a way to match the remarklint configuration with Markdownlint, but I think leaving it alone is best for now.

---

**Duplication**

> • I noted this above #2310 (comment), so maybe duplication should also be removed

Yep. Unfortunately the javascript duplication engine is not particularly useful for this repo at the moment. The engine emits issues far too often for similar function arguments vs function structure.

---

**Node Security**

> • It doesn't appear that this is available to use as a filter when looking at the issues https://codeclimate.com/github/codeclimate-testing/stylelint/issues, does this not show if there are no security issues or is it a code climate bug? (p.s. I just checked and there are no outstanding NSP issues with stylelint)

NSP issues would show up if the engine detects a vulnerable package. They would be visible under the "Security" category. If you haven't found any outstanding NSP issues, then it seems like the engine result is correct. :ok_hand:

---

> Per my above comments what benefit does analysis of stylelint by code climate bring us?

The stylelint team already uses static analysis to ensure a consistent javascript style by running ESLint during CI runs (which is awesome :+1:).

The biggest benefit right now in using Code Climate over your CI environment for code quality issues comes from a contributor's perspective.

If I open up a PR against stylelint/stylelint today, adding a new rule, the Travis CI build might fail. Based on that status alone, a number of things could be wrong:

- The test suite failed
- A linting tool emitted issues
- The build errored (bad config / networking issues)

Let's say that I followed all of the Travis CI links, and discover that ESLint emitted an issue. I don't get much information from the standard ESLint output in a CI environment. I'd need to open up the right file on my system, at the right line, and figure out what I did wrong and how to fix it.

With Code Climate, contributors have much better visibility into issues via Code Climate's PR commit status line, PR compare page, and browser extension.

As a contributor, I can easily see if I've introduced any code quality issues. If I have, issues are shown along with the appropriate source code and with additional inline documentation from the tool explaining what the issue is, why this is an issue, and how to fix it.

> An interesting side effect of this PR is that jest -coverage now works, which I'm presuming would now allow us to resolve #2174

Yep. Enabling Jest coverage reporting and sending that report to Code Climate is pretty simple.

**Comment by davidtheclark** [comment]
@dblandin Is there any CodeClimate support for Flow?

@stylelint/core: In #2062 I said that I did not want to add a CodeClimate badge because I don't think its analysis was helpful for us as an indicator of our code quality. I still don't want that ... but @dblandin's argument above that CodeClimate would be useful for contributors is a totally different one. If anybody would really like to try this out and monitoring how it goes — anybody wants to champion the feature — then I have no objections to the experiment. Anybody interested?

**Comment by dblandin** [comment]
> @dblandin Is there any CodeClimate support for Flow?

@davidtheclark There is no flow engine for Code Climate yet. After playing with it a bit, I couldn't get it to work in a completely static way. Seems like it might require dependencies to be installed first which isn't supported via Code Climate today.

> In #2062 I said that I did not want to add a CodeClimate badge because I don't think its analysis was helpful for us as an indicator of our code quality.

As I would expect, by only running the ESLint and Node Security engines, your project gets a 4.0 from Code Climate, which is probably a more accurate representation of your code quality preferences. The test coverage badge is reporting 96% coverage. If those numbers still aren't helpful, I'm happy to remove the badge additions from this PR.

Sounds like the remaining steps would be to:

- [ ] Decide on the CC badges
- [ ] Use the CC repo token for stylelint/stylelint (not my fork)
- [ ] Configure PR integration

**Comment by dblandin** [comment]
@ntwb Any thoughts on the proposed revisions?

**Comment by ntwb** [comment]
@dblandin Profound apologies for the very late reply.

We're going to pass on adding Code Climate, we'll continue to use Travis CI for ESLint and Jest tests, and have Travis CI forward test coverage to coveralls once the Jest memory issue is sorted.

**Comment by dblandin** [comment]
@ntwb Sounds good! Thanks for the review.
